exercises/src/main/scala/fpinscala/state/State.scala

```diff
@@ -30,32 +29,106 @@ object RNG {
       (f(a), rng2)
     }
 
-  def nonNegativeInt(rng: RNG): (Int, RNG) = ???
+  /*
+  This is skewed slightly because it will map both -1 and 0 to 0.
+  */
+  def nonNegativeInt(rng: RNG): (Int, RNG) = {
+    val (x, rng2) = rng.nextInt
+
+    (if (x < 0) -(x + 1) else x, rng2)
+  }
+
+  /*
+  Repeatedly generating random numbers until we get one that fits the
+  properties we want is less biased than mapping multiple numbers from
+  the domain into one number in the range.
+  */
+  @annotation.tailrec
+  def double(rng: RNG): (Double, RNG) = {
+    val zMax = Int.MaxValue
+    val (x, rng2) = nonNegativeInt(rng)
+
+    if (x == zMax) double(rng2) else (x.toDouble / zMax, rng2)
+  }
+
+  def intDouble(rng: RNG): ((Int,Double), RNG) = {
+    val (i, rng2) = rng.nextInt
+    val (d, rng3) = double(rng2)
+
+    i -> d -> rng3
+  }
+
+  def doubleInt(rng: RNG): ((Double,Int), RNG) = {
+    val ((i, d), rng2) = intDouble(rng)
+
+    d -> i -> rng2
+  }
+
+  def double3(rng: RNG): ((Double,Double,Double), RNG) = {
+    val (d1, rng2) = double(rng)
+    val (d2, rng3) = double(rng2)
+    val (d3, rng4) = double(rng3)
+
+    (d1, d2, d3) -> rng4
+  }
+
+  def ints(count: Int)(rng: RNG): (List[Int], RNG) = {
```

Comment:
I didn't consider using a fold; I felt it would be more natural to use an unfold or an iterate function to express this.

About the re-implementation, you're absolutely right--I'll do that now.